hyperspace/core/README.md

```diff
@@ -0,0 +1,91 @@
+# Hyperspace Relayer
+
+## Architecture
+The relayer is designed to be:
+
+**1. Statelessness**  
+The relayer is designed to be stateless and does not perform any form of caching. The relayer therefore relies heavily on  
+the nodes it's connected to for sourcing data, this design choice eliminates the chances of bugs that could come  
+from cache invalidation problems.
+
+**2. Event Driven**  
+The relayer follows an event driven model, it remains idle if no finality events are received from either chain.
+
+## Relayer Loop
+
+The relayer has a single entry point, which is the [`relay`](/hyperspace/core/src/lib.rs#L20) function, this function takes two [`Chain`](/hyperspace/primitives/src/lib.rs#L346) implementations  
+alongside optional metric handlers and starts the relayer loop.  
+
+The relayer loops awaits finality events from the finality subscription of the chain handlers.  
+Whenever a finality event is received, the latest ibc events are queried using `query_latest_ibc_events`.  
+These events are then parsed into appropriate messages using the `parse_events` function.  
+
+The `parse_events` function internally calls `query_ready_and_timed_out_packets` which queries a chain and  
+produces all packet messages that have passed the connection delay check.
+It also returns timed out packet messages that have passed the connection delay check.  
+
+## Using the relayer
+
+Using the relayer requires having a [`Chain`](/hyperspace/primitives/src/lib.rs#L346) implementation for the chain types  
+that packets would be relayed between.
+
+```rust
+    // Naive example of how to use the relayer
+    pub struct ChainA { ... }
+
+    impl IbcProvider for ChainA { ... }
+    impl KeyProvider for ChainA { ... }
+    impl Chain for ChainA { ... }
+
+    pub struct ChainB { ... }
+
+    impl IbcProvider for ChainB { ... }
+    impl KeyProvider for ChainB { ... }
+    impl Chain for ChainB { ... }
+
+    async fn main() -> Result<(), anyhow::Error>{
+        let chain_a = ChainA::default();
+        let chain_b = ChainB::default();
+        relay(chain_a, chain_b, None, None).await?;
+        Ok(())
+    }
```

Comment:
```suggestion
```rust
    // Naive example of how to use the relayer
    pub struct ChainA { ... }

    impl hyperspace_primitives::IbcProvider for ChainA { ... }
    impl hyperspace_primitives::KeyProvider for ChainA { ... }
    impl hyperspace_primitives::Chain for ChainA { ... }

    pub struct ChainB { ... }

    impl hyperspace_primitives::IbcProvider for ChainB { ... }
    impl hyperspace_primitives::KeyProvider for ChainB { ... }
    impl hyperspace_primitives::Chain for ChainB { ... }

    async fn main() -> Result<(), anyhow::Error>{
        let chain_a = ChainA::default();
        let chain_b = ChainB::default();
        hyperspace_core::relay(chain_a, chain_b, None, None).await?;
        Ok(())
    }
```